### PR TITLE
oc debug use MostAccuratePodTemplateForObject

### PR DESCRIPTION
**Pull request by wanghaoran1988**
fix issue: https://github.com/openshift/origin/issues/14915
I created a [pr](https://github.com/kubernetes/kubernetes/pull/48868) in upstream to add MostAccuratePodTemplateForObject in the objectMappingFactory, but I am not sure whether it's deserved there, for this, I add this func to our factory, and now it works both for a 0 replica deployment and the pipe scenario:

```
➜  origin git:(oc_debug) oc get deployment                       
NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
nginx     0         0         0            0           19h
➜  origin git:(oc_debug) _output/local/bin/linux/amd64/oc debug deploy/nginx
Debugging with pod/nginx-debug, original command: nginx -g daemon off;
Waiting for pod to start ...
Pod IP: 172.17.0.2
If you don't see a command prompt, try pressing enter.

# exit

Removing debug pod ...
➜  origin git:(oc_debug) _output/local/bin/linux/amd64/oc get deployment/nginx -o json |_output/local/bin/linux/amd64/oc debug -f - -- echo hello debug !
Debugging with pod/nginx-debug, original command: nginx -g daemon off;
Waiting for pod to start ...
hello debug !

Removing debug pod ...

```

/cc @smarterclayton @fabianofranz

### CODE REVIEW

**Comment by fabianofranz** [comment]
[test]

**Comment by wanghaoran1988** [comment]
@soltysh do you have bandwidth review this ?

File: pkg/cmd/util/clientcmd/factory.go [hunk 1]
```diff
@@ -353,6 +348,114 @@ func (f *Factory) PodForResource(resource string, timeout time.Duration) (string
 	}
 }
 
+func (f *Factory) MostAccuratePodTemplateForObject(object runtime.Object) (*api.PodTemplateSpec, error) {
```
Comment (soltysh) [inline pkg/cmd/util/clientcmd/factory.go]: godoc

File: pkg/cmd/util/clientcmd/factory.go [hunk 2]
```diff
@@ -418,3 +521,30 @@ func (g groupResourcesByName) Less(i, j int) bool {
 	return g[i].Group < g[j].Group
 }
 func (g groupResourcesByName) Swap(i, j int) { g[i], g[j] = g[j], g[i] }
+
+// GetAllPods returns all pods matching the namespace and label selector
+func GetAllPods(client coreclient.PodsGetter, namespace string, selector labels.Selector, sortBy func([]*v1.Pod) sort.Interface) ([]*api.Pod, error) {
```
Comment (soltysh) [inline pkg/cmd/util/clientcmd/factory.go]: Wouldn't it be better to modify upstream's `GetFirstPod` to watch for pod or not?

File: pkg/cmd/util/clientcmd/factory.go [hunk 3]
```diff
@@ -353,6 +348,114 @@ func (f *Factory) PodForResource(resource string, timeout time.Duration) (string
...
+	var selector labels.Selector
+	var namespace string
+	switch t := object.(type) {
+	case *extensions.ReplicaSet:
+		namespace = t.Namespace
+		selector = labels.SelectorFromSet(t.Spec.Selector.MatchLabels)
+		if err != nil {
+			return &t.Spec.Template, fmt.Errorf("invalid label selector: %v", err)
+		}
+		pods, err := GetAllPods(clientset.Core(), namespace, selector, sortBy)
```
Comment (soltysh) [inline pkg/cmd/util/clientcmd/factory.go]: You're copy&pasting this piece of code. Move it to outside for the case body, similarly to how upstream does `AttachablePodForObject`.

File: pkg/cmd/util/clientcmd/factory.go [hunk 4]
```diff
@@ -207,23 +211,14 @@ func (f *Factory) ApproximatePodTemplateForObject(object runtime.Object) (*api.P
 		return fallback, nil
 
 	default:
-		pod, err := f.AttachablePodForObject(object, 1*time.Minute)
```
Comment (soltysh) [inline pkg/cmd/util/clientcmd/factory.go]: Why this minute timeout doesn't have any effect, in the original issue Clayton pointed out?

**Comment by soltysh** [comment]
Additionally, drop the commits modifying release-related data. That's not needed.

File: pkg/cmd/util/clientcmd/factory.go (hunk 4 above)
Comment (wanghaoran1988) [inline pkg/cmd/util/clientcmd/factory.go]: A 0 replica deployment will time out here and cannot get the template

File: pkg/cmd/util/clientcmd/factory.go (hunk 2 above)
Comment (wanghaoran1988) [inline pkg/cmd/util/clientcmd/factory.go]: Sounds good

Comment (wanghaoran1988) [inline pkg/cmd/util/clientcmd/factory.go]: Do you mean add an "wait" param and  modify like this? 
```
func GetFirstPod(client coreclient.PodsGetter, namespace string, selector labels.Selector, timeout time.Duration, wait bool, sortBy func([]*v1.Pod) sort.Interface) 
```

**Comment by wanghaoran1988** [comment]
@soltysh Do you think we need upstream the MostAccuratePodTemplateForObject ? If so I will update this https://github.com/kubernetes/kubernetes/pull/48868

**Comment by smarterclayton** [comment]
Yes, we should upstream it (other objects like Jobs would benefit from it).

**Pull request closed without merging**